Userland/Libraries/LibWeb/HTML/CloseWatcherManager.cpp

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2024, the Ladybird developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/TypeCasts.h>
+#include <LibWeb/DOM/Document.h>
+#include <LibWeb/DOM/EventDispatcher.h>
+#include <LibWeb/DOM/IDLEventListener.h>
+#include <LibWeb/HTML/CloseWatcher.h>
+#include <LibWeb/HTML/CloseWatcherManager.h>
+#include <LibWeb/HTML/Window.h>
+
+namespace Web::HTML {
+
+JS_DEFINE_ALLOCATOR(CloseWatcherManager);
+
+JS::NonnullGCPtr<CloseWatcherManager> CloseWatcherManager::create(JS::Realm& realm)
+{
+    return realm.heap().allocate<CloseWatcherManager>(realm, realm);
+}
+
+CloseWatcherManager::CloseWatcherManager(JS::Realm& realm)
+    : PlatformObject(realm)
+{
+}
+
+void CloseWatcherManager::add(JS::NonnullGCPtr<CloseWatcher> close_watcher)
+{
+    // If manager's groups's size is less than manager's allowed number of groups
+    if (m_groups.size() < m_allowed_number_of_groups) {
+        // then append « closeWatcher » to manager's groups.
+        Vector<JS::NonnullGCPtr<CloseWatcher>> new_group;
+        new_group.append(close_watcher);
+        m_groups.append(move(new_group));
+    } else {
+        // Assert: manager's groups's size is at least 1 in this branch, since manager's allowed number of groups is always at least 1.
+        VERIFY(!m_groups.is_empty());
+        // Append closeWatcher to manager's groups's last item.
+        m_groups.last().append(close_watcher);
+    }
+
+    // Set manager's next user interaction allows a new group to true.
+    m_next_user_interaction_allows_a_new_group = true;
+}
+
+void CloseWatcherManager::remove(CloseWatcher const& close_watcher)
+{
+    // 2. For each group of manager's groups: remove closeWatcher from group
+    for (auto& group : m_groups) {
+        group.remove_first_matching([&close_watcher](JS::NonnullGCPtr<CloseWatcher>& entry) {
+            return entry.ptr() == &close_watcher;
+        });
+        // 3. Remove any item from manager's group that is empty
+        if (group.is_empty()) {
+            m_groups.remove_first_matching([&group](Vector<JS::NonnullGCPtr<CloseWatcher>> const& entry) {
+                return entry == group;
+            });
+        }
+    }
+}
+
+// https://html.spec.whatwg.org/multipage/interaction.html#process-close-watchers
+bool CloseWatcherManager::process_close_watchers()
+{
+    // 1. Let processedACloseWatcher be false.
+    bool processed_a_close_watcher = false;
+    // 2. If window's close watcher manager's groups is not empty:
+    if (!m_groups.is_empty()) {
+        // 2.1 Let group be the last item in window's close watcher manager's groups.
+        auto& group = m_groups.last();
+        Vector<JS::NonnullGCPtr<CloseWatcher>> group_copy(group);
```

Comment:
Note that `Vector` contents are not visible to the garbage collector.
In other words, this vector does not protect the CloseWatchers from being GC'd if nothing else references them.

Since `request_close()` may dispatch an event and run arbitrary JS, this could be a concern!

You can use `JS::MarkedVector` to get a Vector-compatible container that guarantees GC protection.

Reply:
Also changed CloseWatcherManager::add I assume that should be MarkedVector too?